src/main/lrsql/util/oidc.clj

```diff
@@ -1,19 +1,187 @@
+            [lrsql.spec.config :as config]
+            [lrsql.spec.oidc :as oidc]
+            [lrsql.util.auth :as auth]
+            [selmer.parser :as selm-parser]
+            [xapi-schema.spec :as xs])
+  (:import [java.io File]))
+
+(defn- select-config
+  [config]
+  (select-keys
+   config
+   [:oidc-issuer
+    :oidc-config
+    :jwks-uri]))
 
 (s/fdef get-configuration
-  :args (s/cat :webserver ::config/webserver)
+  :args (s/cat :config ::config/webserver)
   :ret (s/nilable map?))
 
 (defn get-configuration
   "Given webserver config, return an openid configuration if one is specified
-  via :openid-issuer or :openid-config."
-  [{:keys [openid-issuer
-           openid-config]}]
-  (when-let [config-uri (or openid-config
-                            (and openid-issuer
-                                 (disco/issuer->config-uri openid-issuer)))]
-    (disco/get-openid-config config-uri)))
+  via :oidc-issuer or :oidc-config."
+  [{:keys               [oidc-issuer
+           oidc-config] :as config}]
+  (try
+    (when-let [config-uri (or oidc-config
+                              (and oidc-issuer
+                                   (disco/issuer->config-uri oidc-issuer)))]
+      (disco/get-openid-config config-uri))
+    (catch AssertionError ae
+      (ex-info "Invalid OIDC Config"
+               {:type        ::invalid-config
+                :oidc-config (select-config config)}
+               ae))))
+
+(s/fdef resource-interceptors
+  :args (s/cat :config ::config/webserver)
+  :ret (s/every i/interceptor?))
+
+(defn resource-interceptors
+  "Given a webserver config, return a (possibly empty) vector of interceptors.
+  Interceptors will enable token auth against OIDC"
+  [{:keys [jwks-uri] :as config}]
+  (try
+    (if-let [jwks-uri (or jwks-uri
+                          (some-> config
+                                  get-configuration
+                                  (get "jwks_uri")))]
+      (let [keyset-cache (atom (jwt/get-keyset jwks-uri))]
+        [;; Decode/Unsign tokens
+         (oidc-i/decode-interceptor
+          (fn [_]
+            (fn [kid]
+              (get @keyset-cache kid
+                   ;; If kid is not found in the keyset, attempt refresh and try
+                   ;; again
+                   (get (reset! keyset-cache (jwt/get-keyset jwks-uri))
+                        kid))))
+          :required? false)
+         ;; This is a vector in case we need additional interceptors. At present
+         ;; we do not.
+         ])
+      ;; If no config, don't return any
+      [])
+    (catch Exception ex
+      (throw
+       (ex-info
+        "OIDC Initialization Failure"
+        {:type        ::init-failure
+         :oidc-config (select-config config)}
+        ex)))))
+
+(s/fdef parse-scope-claim
+  :args (s/cat :scope-str string?)
+  :ret (s/every ::auth/scope))
+
+(defn parse-scope-claim
+  [scope-str]
+  (keep auth/scope-str->kw
+        (cs/split scope-str #"\s")))
```

Comment:
This might actually get more complex: Currently the scope names are very general (ex. `all`) and we're probably going to want to allow a configurable prefix as many IDP scenarios consider scopes global, probably leading to a need for tests, in which case it could not be private.

Reply:
In this case there should be a comment explaining why this helper function should be public.

Reply:
I think I'll just implement the rest of the scope prefix feature, we're almost certainly gonna need it anyhow, and it makes this function less trivial

Reply:
Implemented the scope prefix here: https://github.com/yetanalytics/lrsql/pull/205/commits/0002110e9983fcad6d220ae001185494b8ccd817